packages/core/core/src/Asset.js

```diff
@@ -46,63 +59,175 @@ type SerializedOptions = {|
     this.type = options.type;
-    this.code = options.code || (options.output ? options.output.code : '');
+    this.content = options.content || '';
+    this.contentKey = options.contentKey;
     this.ast = options.ast || null;
+    this.map = options.map;
     this.dependencies = options.dependencies
       ? new Map(options.dependencies)
       : new Map();
     this.connectedFiles = options.connectedFiles
       ? new Map(options.connectedFiles)
       : new Map();
-    this.output = options.output || {code: this.code};
     this.outputHash = options.outputHash || '';
     this.env = options.env;
     this.meta = options.meta || {};
-    this.stats = options.stats || {
-      time: 0,
-      size: this.output.code.length
-    };
+    this.stats = options.stats;
   }
 
   serialize(): SerializedOptions {
-    // Exclude `code` and `ast` from cache
+    // Exclude `code`, `map`, and `ast` from cache
     return {
       id: this.id,
       hash: this.hash,
       filePath: this.filePath,
       type: this.type,
       dependencies: Array.from(this.dependencies),
       connectedFiles: Array.from(this.connectedFiles),
-      output: this.output,
+      isIsolated: this.isIsolated,
       outputHash: this.outputHash,
       env: this.env,
       meta: this.meta,
-      stats: this.stats
+      stats: this.stats,
+      contentKey: this.contentKey
     };
   }
 
+  /*
+   * Prepares the asset for being serialized to the cache by commiting its
+   * content and map of the asset to the cache.
+   */
+  async commit(): Promise<void> {
+    this.ast = null;
+
+    let contentStream = this.getStream();
+    if (
+      // $FlowFixMe
+      typeof contentStream.bytesRead === 'number' &&
+      contentStream.bytesRead > 0
+    ) {
+      throw new Error(
+        'Stream has already been read. This may happen if a plugin reads from a stream and does not replace it.'
+      );
+    }
+
+    let size = 0;
+    let hash = crypto.createHash('md5');
+
+    // Since we can only read from the stream once, compute the content length
+    // and hash while it's being written to the cache.
+    let [contentKey, mapKey] = await Promise.all([
+      Cache.setStream(
+        this.generateCacheKey('content'),
+        contentStream.pipe(
+          new TapStream(buf => {
+            size += buf.length;
+            hash.update(buf);
+          })
+        )
+      ),
+      this.map == null
+        ? Promise.resolve()
+        : Cache.set(this.generateCacheKey('map'), this.map)
+    ]);
+    this.contentKey = contentKey;
+    this.mapKey = mapKey;
+    this.stats.size = size;
+    this.outputHash = hash.digest('hex');
+  }
+
+  async getCode(): Promise<string> {
+    if (this.contentKey) {
+      this.content = Cache.getStream(this.contentKey);
+    }
+
+    if (typeof this.content === 'string' || this.content instanceof Buffer) {
+      return this.content.toString();
+    }
+
+    this.content = (await bufferStream(this.content)).toString();
```

Comment:
maybe we should just work with binary buffers or streams internally, and always convert to a string whenever `code` is requested? Would simplify some of this.

Reply:
I wonder about the perf implications of always reading from a stream vs just using `readFile`. Node might optimize `readFile` internally, not sure.